### PR TITLE
Ft users edit request 167735062

### PR DESCRIPTION
**Pull request by cleave3**
**What does this PR do?**

  If merged, this PR will add endpoint for user to edit their pending trip requests

  **Description of Task to be completed?**
  - Create request model.
  - Add sequelize seeder for trip request.
  - Create endpoint for editing trip request.
  - Add validation for user input.
  - Add middleware for verifying user.

  **How should this be manually tested?**

  - Rebase branch
 - On the terminal run npm run db:migration and npm run seed.
 - On postman add token to headers.
 - Then, send a patch request to /api/v1/request/:id . use 11 as id for pending request, 22 for accepted and 33 for rejected request.
 - Required field are : source, destination, tripType, travelDate, returnDate, reason, accommodation

  **Any background context you want to provide?**

  Logged in user should be able to edit request that are still open

  **What are the relevant stories?**

 #167735062

  **Screenshots (if appropriate)**

 
image
image

  **Questions:**

None

### CODE REVIEW

File: src/controllers/requestController.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * Class for Requests
+ */
+export default class RequestController {
+/**
+ * Update a pending trip request
+ * @param {*} req - requset object
+ * @param {*} res - response object
+ * @return {json} - json
+ */
+  static async updateRequest(req, res) {
```
Comment (Annmary12) [inline src/controllers/requestController.js]: Indent the jsdoc correctly.
res and req response is an `object` instead of `*`

File: src/controllers/requestController.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    const data = req.body;
+    const { id } = req.params;
+    try {
+      if (data.tripType !== 'return') data.returnDate = null;
+
+      const result = await RequestService.updateRequest(Number(id), data);
+
+      res.status(200).json({ status: 'success', data: result });
```
Comment (Annmary12) [inline src/controllers/requestController.js]: check the util folder, there is a helper function for this `res.status(200).json({ status: 'success', data: result });`, please ultilize that and adjust all your code

File: src/middlewares/auth.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import jwt from 'jsonwebtoken';
+import dotenv from 'dotenv';
+
+dotenv.config();
+const secret = process.env.JWTSECRET;
+
+const verifyUser = (req, res, next) => {
+  const { token } = req.headers;
+  try {
```
Comment (Annmary12) [inline src/middlewares/auth.js:10]: Please, go through your code and ensure that anytime you declare a variable or sets of variables, leave a line.
Example, leave a line after `const { token } = req.headers;`

File: src/middlewares/auth.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    if (!token) {
+      return res.status(403).json({
+        status: 403,
+        error: 'Authentication failed, please login',
+      });
+    }
+    const decoded = jwt.verify(token, secret);
+    if (decoded) {
```
Comment (Annmary12) [inline src/middlewares/auth.js:19]: What happens if the token has expired?

File: src/migrations/20190828145652-create-request.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    createdAt: {
+      allowNull: false,
+      type: Sequelize.DATE
+    },
+    updatedAt: {
+      allowNull: false,
+      type: Sequelize.DATE
+    }
+  }),
+  down: (queryInterface) => queryInterface.dropTable('Requests')
```
[the remark this answers is not in the document]
Comment (Annmary12) [reply src/migrations/20190828145652-create-request.js]: Please attend to hound errors

File: src/routes/index.js [hunk 6]
```diff
@@ -1,9 +1,11 @@
 import { Router } from 'express';
 import authRoute from './auth';
+import requestRoute from './request';
 
 const router = Router();
 
 
 router.use('/auth', authRoute);
+router.use(requestRoute);
```
Comment (Annmary12) [inline src/routes/index.js]: Please rename the name of the route, instead of `requestRoute`, you can use `/requests`

File: src/routes/request.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { Router } from 'express';
+import requestController from '../controllers/requestController';
+import { requestSchema } from '../validation/schemas';
+import validate from '../validation/validator';
+import verifyUser from '../middlewares/auth';
+
+const router = Router();
+
+router.patch('/request/:id', validate(requestSchema, 'body'), verifyUser, requestController.updateRequest);
```
Comment (Annmary12) [inline src/routes/request.js]: Since the root route will be `requests`, this route should be `/:id` not `/request/:id`

Also try and apply destructuring in your code. 

Example
Instead of having `requestController.updateRequest`

you can destructure it up by:
```
const { update } = requestController;

router.patch(':id', validate(requestSchema, 'body'), verifyUser, update);

```

You might not necessarily need to add `Request` to the name of the functions in the request controller.

File: src/services/requestService.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import { Request } from '../models';
+
+/**
+ * Request service class
+ */
+export default class RequestService {
+  /**
+  *update trip rquest
+  * @param {Number} id - request id
```
Comment (Annmary12) [inline src/services/requestService.js]: number `N` should be small letter

File: src/test/auth.request.spec.js [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+describe('TRIP REQUEST ROUTE', () => {
+  describe('Edit Request', () => {
+    it('should update an open trip request when user is logged in and required details are provided', async () => {
+      const res = await chai.request(app)
+        .patch(`${URL_PREFIX}/request/1`)
+        .set('token', userToken)
+        .send(requestDetails);
+
+      res.should.have.status(200);
+      res.body.should.be.an('object');
```
Comment (Annmary12) [inline src/test/auth.request.spec.js:73]: Please, let's assert that a message is equal to the expected message not just expecting an object

File: src/seeders/20190828170312-demo-request.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  accommodation: 'Lorem ipsum',
+  status,
+  userId: 1,
+  createdAt: new Date(),
+  updatedAt: new Date()
+});
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.bulkInsert('Requests', [
+    demoRequest(1, 'pending'),
+    demoRequest(2, 'accepted')
```
Comment (Annmary12) [inline src/seeders/20190828170312-demo-request.js]: Add `rejected` as part of the status

File: src/utils/helpers.js [hunk 11]
```diff
@@ -61,7 +61,7 @@ export default class Helper {
    */
   static genToken(payloader) {
     const secret = process.env.JWTSECRET;
-    const token = jwt.sign(payloader, secret, { expiresIn: '1hr' });
+    const token = jwt.sign(payloader, secret, { expiresIn: '12h' });
```
Comment (Annmary12) [inline src/utils/helpers.js]: Why did you change it to expire in `12h`?

**Comment by Annmary12** [comment]
Fix conflict on your PR and add swagger docs too

File: src/middlewares/auth.js (hunk 4 above)
Comment (cleave3) [reply src/middlewares/auth.js:19]: The catch block handles that by throwing error access denied

**Comment by Annmary12** [comment]
@cleave3  I could edit a request that does not belong to me. Please fix that

**Comment by Annmary12** [comment]
@cleave3, I noticed that before your test could run, I have to run `npm run seed`, Ideally, it shouldn't work that way because you are not actually testing at that point. In your beforeScript, you can create a request before you can write test to edit it. Or you can reach out to @SamuelKoroh , he created a test helper function that you can use to create a `request` before you can edit.
Also, reach out to who is working on creating a request so that you can work hand in hand on how you can create a request.

**Comment by Annmary12** [comment]
Screenshot 2019-09-02 at 8 07 10 PM


Lastly, this test is failing on my system
